### PR TITLE
Make tensorflow-serving-cluster a helm package

### PR DESCRIPTION
**Pull request by snormore**
[helm](https://github.com/kubernetes/helm) is a package manager for Kubernetes.

There's no need for a Python package with a cli interface, the proxy server code is distributed in the Docker image, helm is the cli.

This PR is still a WIP, but you should be able to get a cluster running in development using the README.

----

[Build Status](https://travis-ci.org/triagemd/tensorflow-serving-cluster)

# Getting started

Install [Docker Toolbox](https://www.docker.com/products/docker-toolbox) and [MiniKube](https://kubernetes.io/docs/tasks/tools/install-minikube/).

### Start up the cluster

```bash
minikube start
helm init
script/distribute-dev
helm install --name acme helm/tensorflow-serving-cluster \
  --set cluster.config.b64=$(base64 cluster.yml) \
  --set image.tag=dev
# ...
script/distribute-dev
helm upgrade acme helm/tensorflow-serving-cluster \
  --set cluster.config.b64=$(base64 cluster.yml) \
  --set image.tag=dev
```

### Kubernetes dashboard

```bash
minikube dashboard
```

### Tear down and clean up

```bash
helm delete acme
minikube stop
```

### Start up the cluster

```bash
gcloud container cluster create acme-serving
helm init
helm install --name acme helm/tensorflow-serving-cluster \
  --set cluster.config.b64=$(base64 cluster.yml)
helm install --name lb stable/nginx-ingress
# ... grab the external IP of the nginx service, add/update DNS A record to point to it, wait until the domain is reachable
helm install --name ssl-certs stable/kube-lego \
  --set config.LEGO_EMAIL=steven@triage.com \
  --set config.LEGO_URL=https://acme-v01.api.letsencrypt.org/directory
helm upgrade acme helm/tensorflow-serving-cluster \
  --set service.tls=true \
  --set cluster.config.b64=$(base64 cluster.yml)
```

### Query the cluster

```bash
time curl --user foo:secret $(minikube service acme-tensorflow-serving-cluster --url)/imagenet_mobilenet_v1/classify?image=https://image.ibb.co/nu62ba/cat.jpg
time curl --user foo:secret $(minikube service acme-tensorflow-serving-cluster --url)/imagenet_inception_v3/classify?image=https://image.ibb.co/nu62ba/cat.jpg
time curl --user foo:secret $(minikube service acme-tensorflow-serving-cluster --url)/imagenet_resnet50/classify?image=https://image.ibb.co/nu62ba/cat.jpg
```

### Kubernetes dashboard

```bash
kubectl proxy
open http://localhost:8001/ui
```

### Tear down and clean up

```bash
helm delete ssl-certs --purge
helm delete lb --purge
helm delete acme --purge
```

Tear down all deployed model servers while leaving the proxy manager up:

```bash
kubectl delete --all deployments,services --namespace=tensorflow-serving-cluster-models
```

### Exec into a running container

```bash
kubectl exec -it <pod> -- /bin/bash
```

# Live development and testing

Install [Docker Toolbox](https://www.docker.com/products/docker-toolbox)

Install Python 3 using `virtualenv .venv -p python3 && source .venv/bin/activate`, or just `source script/env`.

## Docker Compose

 - `script/up`
 - `script/test`
 - `time curl localhost:5000/imagenet_mobilenet_v1/classify?image=https://image.ibb.co/nu62ba/cat.jpg`
 - `time curl localhost:5000/imagenet_inception_v3/classify?image=https://image.ibb.co/nu62ba/cat.jpg`
 - `time curl localhost:5000/imagenet_resnet50/classify?image=https://image.ibb.co/nu62ba/cat.jpg`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -9,83 +9,107 @@ A cluster of [tensorflow-serving](https://www.tensorflow.org/serving/) model ser
...
+## Production (GKE)
 
-# Getting started
+### Start up the cluster
 
-... on a development and testing environment.
+```bash
+gcloud container cluster create acme-serving
+helm init
+helm install --name acme helm/tensorflow-serving-cluster \
```
Comment (snormore) [inline README.md]: This needs to reference a `tgz` URL

File: docker/python3.6-tensorflow-optimized/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,23 @@
+FROM python:3.6
+
+# Instal bazel
```
Comment (snormore) [inline docker/python3.6-tensorflow-optimized/Dockerfile]: This will be moved out into it's own PR.

Comment (stephensolis) [inline docker/python3.6-tensorflow-optimized/Dockerfile]: Looks great though 👍

File: docker/python3.6-tensorflow-optimized/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    update-alternatives --config java && \
+    "deb http://storage.googleapis.com/bazel-apt stable jdk1.8" | tee /etc/apt/sources.list.d/bazel.list && \
+    https://storage.googleapis.com/bazel-apt/doc/apt-key.pub.gpg | apt-key add - && \
+    apt-get update && \
+    apt-get -y install bazel
+
+# Build optimized tensorflow and install
+RUN git clone https://github.com/tensorflow/tensorflow && \
+    cd tensorflow && \
+    git reset --hard a23f5d7 && \
```
Comment (stephensolis) [inline docker/python3.6-tensorflow-optimized/Dockerfile]: Is this supposed to be a tensorflow release? How about eg. `git checkout tags/v1.2.1` instead?

File: docker/python3.6-tensorflow-optimized/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    ./configure && \
+    bazel build -c opt --copt=-mavx --copt=-mavx2 --copt=-mfma --copt=-mfpmath=both --copt=-msse4.2 --copt=-msse4.1 --copt=-msse3-k //tensorflow/tools/pip_package:build_pip_package && \
+    bazel-bin/tensorflow/tools/pip_package/build_pip_package /tmp/tensorflow_pkg && \
```
Comment (stephensolis) [inline docker/python3.6-tensorflow-optimized/Dockerfile]: Don't know how much I like doing this in a Dockerfile. How about we set up a CI server and build wheels on every Tensorflow release, we can stash them on our CDN or even publicly on Bintray or something.

File: docker/python3.6-tensorflow-optimized/Dockerfile (hunk 3 above)
Comment (snormore) [inline docker/python3.6-tensorflow-optimized/Dockerfile]: Yup that sha doesn't make sense

File: docker/python3.6-tensorflow-optimized/Dockerfile (hunk 4 above)
Comment (snormore) [inline docker/python3.6-tensorflow-optimized/Dockerfile]: Yea I like that 👍

File: helm/tensorflow-serving-cluster/templates/NOTES.txt [hunk 5]
```diff
@@ -0,0 +1,15 @@
+1. Get the application URL by running these commands:
```
Comment (snormore) [inline helm/tensorflow-serving-cluster/templates/NOTES.txt]: This needs to be updated, it's still the generated default.

File: tensorflow_serving_cluster/proxy_app.py [hunk 6]
```diff
@@ -88,12 +101,19 @@ def model(model_name):
...
-        if os.environ.get('SERVING_LOCALHOST'):
+        if os.environ.get('SERVING_LOCAL'):
             serving_host = 'localhost'
         else:
             serving_host = model_name.replace('_', '-')
+            serving_host = '%s.tensorflow-serving-cluster-models.svc.cluster.local' % (serving_host, )
+
+        if not hostname_resolves(serving_host):
+            print('Creating k8s service for %s' % (model_name, ))
+            service = app.cluster.build_model_service(model_name, model_config)
```
Comment (snormore) [inline tensorflow_serving_cluster/proxy_app.py]: This needs better test coverage.

Comment (stephensolis) [inline tensorflow_serving_cluster/proxy_app.py]: I'm not sure how concerned I am about this yet, but AFAIU images would be sent between cluster nodes in plaintext. We're going to need to be very careful about what that does to our HIPAA posture.
